audio-worklet/src/noisegate-audio-worklet.js

```diff
@@ -0,0 +1,194 @@
+
+    let inputChannelData = input.getChannelData(0);
+    let outputChannelData = output.getChannelData(0);
+
+    let envelope = this.detectLevel_(inputChannelData);
+    let weights = this.computeWeights_(envelope);
+
+    for (let j = 0; j < inputChannelData.length; j++) {
+      outputChannelData[j] = weights[j] * inputChannelData[j];
+    }
+  }
+
+  /**
+   * Compute an envelope follower for the signal.
+   * @param {Float32Array} channelData Input channel data.
+   * @return {Float32Array} The level of the signal.
+   */
+  detectLevel_(channelData) {
+    // The signal level is determined by filtering the square of the signal
+    // with exponential smoothing. See
+    // http://www.aes.org/e-lib/browse.cfm?elib=16354 for details.
+    this.envelope_[0] = this.alpha_ * this.previousLevel_ +
+        (1 - this.alpha_) * Math.pow(channelData[0], 2);
+
+    for (let j = 1; j < channelData.length; j++) {
+      this.envelope_[j] = this.alpha_ * this.envelope_[j - 1] +
+          (1 - this.alpha_) * Math.pow(channelData[j], 2);
+    }
+    this.previousLevel_ = this.envelope_[this.envelope_.length - 1];
+
+    return this.envelope_;
+  }
+
+  /**
+   * Computes an array of weights which determines what samples are silenced.
+   * @param {Float32Array} envelope The output from envelope follower.
+   * @return {Float32Array} weights Numbers in the range 0 to 1 set in
+   *                                accordance with the threshold, the envelope,
+   *                                and attack and release.
+   */
+  computeWeights_(envelope) {
+    // When attack or release are 0, the weight changes between 0 and 1
+    // in one step.
+    let attackSteps = 1;
+    let releaseSteps = 1;
+    let attackLossPerStep = 1;
+    let releaseGainPerStep = 1;
+
+    // TODO: Replace this weights-based approach for enabling attack/release
+    // parameters with the method described on page 22 in
+    // "Signal Processing Techniques for Digital Audio Effects".
+
+    // When attack or release are > 0, the associated weight changes between 0
+    // and 1 in the number of steps corresponding to the millisecond attack
+    // or release time parameters.
+    if (this.attack > 0) {
+      attackSteps = Math.ceil(this.sampleRate * this.attack);
+      attackLossPerStep = 1 / attackSteps;
+    }
+    if (this.release > 0) {
+      releaseSteps = Math.ceil(this.sampleRate * this.release);
+      releaseGainPerStep = 1 / releaseSteps;
+    }
+
+    // Compute an array of weights between 0 and 1 which will be multiplied with
+    // the channel depending on if the noise gate is open, attacking, releasing,
+    // or closed.
+    for (let i = 0; i < envelope.length; i++) {
+      // For sine waves, the envelope eventually reaches an average power of
+      // a^2 / 2. Sine waves are therefore scaled back to the original
+      // amplitude, but other waveforms or constant sources can only be
+      // approximated.
+      const scaledEnvelopeValue
```

Comment:
Yes.  In the script processor if you generate garbage, you'll end up messing up the main thread and possibly the audio thread if GC takes too long.

In a worklet, if you generate garbage, you're pretty much guaranteed to glitch audio because GC on Chrome can take 10ms or more.